docs/tutorials/developer-journey/level-5/5.1-vetKeys-tutorial.md

```diff
@@ -0,0 +1,794 @@
+import Nat "mo:base/Nat";
+import Bool "mo:base/Bool";
+import Principal "mo:base/Principal";
+import Result "mo:base/Result";
+import Option "mo:base/Option";
+import Debug "mo:base/Debug";
+import Order "mo:base/Order";
+import Blob "mo:base/Blob";
+import Hex "./utils/Hex";
+
+// Declare a shared actor class and bind the caller and the initializer:
+
+shared({ caller = initializer }) actor class() {
+
+    // Currently, a single canister smart contract is limited to 4 GB of storage due to WebAssembly limitations.
+    // To ensure that your canister does not exceed this limit, it is recommended to restrict memory usage to at most 2 GB, as this example does.
+    // This includes the assumption that up to 2x memory may be needed for data serialization during canister upgrades. 
+    // Therefore, this example dapp can support up to 1,000 users, each storing up to 2 MB of data. 
+    // The data is reserved for storing the notes:
+    //     NOTES_PER_USER = MAX_NOTES_PER_USER x MAX_NOTE_CHARS x (4 bytes per char)
+    //     2 MB = 500 x 1000 x 4 = 2,000,000
+
+    // Next, define the dapp's limits. This is important for security assurance:
+
+    private let MAX_USERS = 1_000;
+    private let MAX_NOTES_PER_USER = 500;
+    private let MAX_NOTE_CHARS = 1000;
+
+    // Define private a type for PrincipalName:
+
+    private type PrincipalName = Text;
+
+    // Define a public type for EncryptedNote:
+
+    public type EncryptedNote = {
+        encrypted_text: Text;
+        id: Nat;
+    };
+
+    // Next, define private fields.
+    // Stable actor fields are automatically retained across canister upgrades. 
+    // You can learn more about upgrades in the Motoko docs: https://internetcomputer.org/docs/current/motoko/main/upgrades/
+
+    // As a design choice, this canister uses globally unique note identifiers for all users.
+    // This identifier is a stable variable, which allows the value to persist across canister upgrades. 
+
+    private stable var nextNoteId: Nat = 1;
+    
+    // Next, the canister stores each user's notes in a separate List. 
+
+    private var notesByUser = Map.HashMap<PrincipalName, List.List<EncryptedNote>>(0, Text.equal, Text.hash);
+    
+    // While accessing data via [notesByUser] is more efficient, this example uses the following stable array as a buffer to preserve user notes across canister upgrades.
+
+    private stable var stable_notesByUser: [(PrincipalName, List.List<EncryptedNote>)] = [];
+
+    // Next, the canister defines a function to return the current number of users.
+    // This function traps if [users_invariant] is violated:
+
+    private func user_count(): Nat {
+        notesByUser.size()
+    };
+
+    // The canister checks that a note identifier is sane. This is needed since Motoko integers are infinite-precision. 
+    // Note: It is recommended to avoid extraneous usage of async functions, hence [user_count]
+
+    private func is_id_sane(id: Nat): Bool {
```

Comment:
question: there seems to be a mix of camelCase and snake_case throughout this file. is the typical format for Motoko programs camelCase? perhaps we should stick with a consistent style throughout the documentation

Reply:
Good catch, thanks Jason! Standardized it to camelCase 🫡